README.md
<img src="https://ufactory.cc/viewpage/img/logo-whole-png.png" width="200">

# pyuarm

This is a uArm Library for Python.
## Requirement

- uArmFirmata - [uArmFirmata V1.5][5bd49e15]

We recommend you to use [FirmwareHelper][328e8ff3], You could flash the firmware to uArm without pain.

## Installation

### Using PYPI
*Recommend*

- pip install pyuarm

### Manual Installation

- Download source code
- python setup.py install

## Getting Start

### Initialize uArm
There are two methods to Initialize uArm.
- Use get_uarm()
```Python
>>> import pyuarm
>>> uarm = pyuarm.get_uarm()
Firmware Version: 1.5.9
```
if no uArm connected, will return `None` and display "There is no uArm Port available"
```Python
>>> import pyuarm
>>> uarm = pyuarm.get_uarm()
There is no uArm port available
>>> print uarm
None
```
- Use uArm()
``` Python
>>> import pyuarm
>>> uarm = pyuarm.uArm()
Firmware Version: 1.5.9
```
If no uArm connected, will raise `NoUArmPortException`.
```Python
>>> uarm = pyuarm.uArm()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyuarm.py", line 114, in __init__
    raise NoUArmPortException()
pyuarm.NoUArmPortException
```
You could turn on Debug Mode
```Python
>>> import pyuarm
>>> uarm = pyuarm.uArm(debug_mode=True)
f0aa23f7
Firmware Version: 1.5.9
```
Define the uArm port
```Python
>>> import pyuarm
>>> uarm = pyuarm.uArm(port='/dev/cu.usbserial-A600CVS1')
Firmware Version: 1.5.9
```

### Movement

```Python
>>> uarm.moveTo(15, -5, 15) ### Absolute coordinate in 2 seconds
>>> uarm.move(5, 5, 5) ### Relative coordinate
```

### Current Positions
```Python
>>> uarm.read_coords() ### Read Current Coordinate
```

### Pump control
```Python
>>> uarm.pump_control(True) ### Pump On
>>> uarm.pump_control(False) ### Pump Off
```

### Others

#### List uArm Ports
```Python
>>> import pyuarm
>>> pyuarm.list_uarms() #List All available uArm Ports
[u'/dev/cu.usbserial-A600CVS1']
```

[5bd49e15]: https://github.com/uArm-Developer/UArmFirmata "uArmFirmata V1.5"
[a3fca171]: https://github.com/uArm-Developer/UArmForArduino "UArmForArduino"
[328e8ff3]: https://github.com/uArm-Developer/FirmwareHelper "FirmwareHelper"


HISTORY.md
## [1.4.1] - 2016-09-26
## Updated - leixiayang
添加了接口被占用的时候提示报错信息

## [1.4.1] - 2016-09-22
## Updated - leixiayang

### Fix

- Move version to __init__.py instead of version.py
- Add __future__ feature for print function
- replace tqdm with progressbar to Compatible with python3
- Fix firmware.hex path issue

## [1.4.0] - 2016-09-22
## Updated - leixiayang
分别修改了 history.md , requirements.txt , __init__.py
pyuarm.py , util.py calibrate.py , firmware_helper.py
list_helper. py , list_uarms.py , miniterm.py 文件内容都进行了不同的修改

## [1.3.32] - 2016-08-31
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- update miniterm, fix some Document issue, add miniterm version

## [1.3.31] - 2016-08-30
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- update miniterm, add gripper control


## [1.3.30] - 2016-08-25
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- update miniterm, read servo angles return 4 angles.


## [1.3.29] - 2016-08-15
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Update calibrate.py, fix hand servo (Null offset)
- change cli to miniterm

## [1.3.28] - 2016-08-15
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Update firmware_helper.py add --upgrade argument, multi ports support
- add cli.py uarm command line
- uarm-calibrate supports multi ports
- move get_uarm() to list_uarms.py
- add get_uarm_port_cli()

## [1.3.27] - 2016-07-08
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Fix calibrate issue when use GUI.

## [1.3.26] - 2016-07-05
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Firmware helper support port argument.
- improve firmware helper in mac os x if not install avrdude.
- add console-script: uarm-firmware, uarm-calibrate, uarm-list.

## [1.3.25] - 2016-07-01
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Move Back to use version.py due to pyinstaller encounter error

## [1.3.24] - 2016-06-30
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Use version.json replace version.py
- Add Support firmware version in version.json, if not Compatible will raise UnsupportFirmwareVersionException
- Add argument for calibrate.py
- Disable Stretch Calibration
- Change Linear Calibration, Increasing moving points
- Merge [frankyjuang](https://github.com/frankyjuang), fix typo

## [1.3.21] - 2016-06-28
### Updated - Alex Tan

### Compatible
- Firmware 1.7.1

### Fix
- rewrite Firmware helper. remove pycurl

## [1.3.20] - 2016-06-28
### Updated - Alex Tan

### Compatible
- Firmware 1.7.1

### Fix
- Increase More delay on calibrate. rewrite Linear Offset points.
- Update requirements.txt


## [1.3.19] - 2016-06-27
### Updated - Alex Tan

### Fix
- Increase More delay on calibrate. rewrite Linear Offset points.
- Update requirements.txt

## [1.3.18] - 2016-06-22
### Updated - Alex Tan

### Fix
- firmware_helper -f won't download the firmware.hex use firmware_helper -d instead

## [1.3.17] - 2016-06-22
### Updated - Alex Tan

### Fix
- firmware_helper remove access token

### To-Do
- Setup Firmware_checker at UFactory Server

## [1.3.16] - 2016-06-22
### Updated - Alex Tan

### Fix
- firmware_helper add APIError check

## [1.3.15] - 2016-06-22
### Updated - Alex Tan

### Fix
- Add enable_hand support Firmware 1.6.1

## [1.3.14] - 2016-06-22
### Updated - Alex Tan

### Fix
- Raise UnkwonFirmwareException when init uArm.

## [1.3.13] - 2016-06-14
### Updated - Alex Tan

### Changes
- Try NetworkError in firmware_helper

## [1.3.12] - 2016-06-14
### Updated - Alex Tan

### Changes
- Compatible with Firmware 1.6.0
- Time.sleep(3) when init Serial Port
- read_servo_angle() when servo_number is none, if servo_number is None, will return all servo angles (SERVO_BOTTOM, SERVO_LEFT, SERVO_RIGHT, SERVO_HAND).
- Update Firmware Release URL

### Fix
- Fix pyuarm.tools.calibrate print format & raw_input

## [1.3.11] - 2016-06-14
### Updated - Alex Tan

### Changes
- include requirements, README.rst

## [1.3.10] - 2016-06-14
### Updated - Alex Tan

### Changes
- include requirements, README.rst

## [1.3.9] - 2016-06-14
### Updated - Alex Tan

### Changes
- add version.py for version variable

## [1.3.8] - 2016-06-14
### Updated - Alex Tan

### Changes
- Replace progressbar with tqdm
- Rewrite firmware_helper into class
- calibrate.py confirm manual trigger in raw_input: y

## [1.3.7] - 2016-06-13
### Updated - Alex Tan

### Changes
- Add requirements.txt

## [1.3.7] - 2016-06-13
### Updated - Alex Tan

### Changes
- Update read_coordinate()
- Update util.py function name getOne7BitBytesFloatArray, getThree7BitBytesFloatArray

## [1.3.6] - 2016-06-13
### Updated - Alex Tan

### Changes
- Update read_coordinate()
- Update util.py function name getOne7BitBytesFloatArray, getThree7BitBytesFloatArray

## [1.3.5] - 2016-06-13
### Updated - Alex Tan

### Changes
- Update Version 1.3.5
- Update Documentation
- Update README.md to README.rst

## [1.3.4] - 2016-06-10
### Updated - Alex Tan

### Changes
- move pyuarm.calibrate to pyuarm.tools.calibrate
- move pyuarm.list_uarms function to pyuarm.tools.list_uarms
- pyuarm.py replace serial_read_byte() with sp.read(1), support debug read message
- firmware_helper.py add NetworkError
- Docstring
- move Exception get_uarm to pyuarm.util
- pyuarm.py set_firmware_version, set_firmata_version to get_firmware_version, get_firmata_version.
- pyuarm.py move_to_opts, move_to_simple, move integrate with one Function move_to
- pyuarm.py changes alert() function to alarm()

## [1.3.3] - 2016-06-08
### Updated - Alex Tan

### Fix
- fix setup.py package_data avrdude folder, MANIFEST.in

## [1.3.2] - 2016-06-08
### Updated - Alex Tan

### Fix
- fix setup.py 'pycurl>=7.43.0', 'certifi>=2016.02.28', 'tqdm>=4.7.2', 'requests>=2.10.0'

## [1.3.2] - 2016-06-08
### Updated - Alex Tan

### Fix
- firmware_helper, improve download progress bar

## [1.3.0] - 2016-06-08
### Updated - Alex Tan

### Changes

pyuarm/util.py
# Firmata

REPORT_FIRMATA_VERSION = 0x79

# parameters
SERVO_BOTTOM = 0
SERVO_LEFT = 1
SERVO_RIGHT = 2
SERVO_HAND = 3

SERVO_BOTTOM_ANALOG_PIN = 2
SERVO_LEFT_ANALOG_PIN = 0
SERVO_RIGHT_ANALOG_PIN = 1
SERVO_HAND_ANALOG_PIN = 3

# communication protocol
START_SYSEX = 0xF0
UARM_CODE = 0xAA
END_SYSEX = 0xF7

READ_ANGLE = 0X10
WRITE_ANGLE = 0X11
READ_COORDS = 0X12
WRITE_COORDS = 0X13
READ_DIGITAL = 0X14
WRITE_DIGITAL = 0X15
READ_ANALOG = 0X16
WRITE_ANALOG = 0X17
READ_EEPROM = 0X1A
WRITE_EEPROM = 0X1B
SERVO_STATUS = 0X1C
PUMP_STATUS = 0X1D
WRITE_STRETCH = 0X1E
WRITE_LEFT_RIGHT_ANGLE = 0X1F
GRIPPER_STATUS = 0X20
READ_SERIAL_NUMBER = 0x21
WRITE_SERIAL_NUMBER = 0x22
REPORT_FIRMWARE_VERSION = 0x23
BUZZER_ALERT = 0x24

CONFIRM_FLAG = 0x80
CALIBRATION_FLAG = 10
CALIBRATION_LINEAR_FLAG = 11
CALIBRATION_SERVO_FLAG = 12
CALIBRATION_STRETCH_FLAG = 13

LINEAR_INTERCEPT_START_ADDRESS = 70
LINEAR_SLOPE_START_ADDRESS = 50
OFFSET_START_ADDRESS = 30
OFFSET_STRETCH_START_ADDRESS = 20

SERIAL_NUMBER_ADDRESS = 100

EEPROM_DATA_TYPE_BYTE = 1
EEPROM_DATA_TYPE_INTEGER = 2
EEPROM_DATA_TYPE_FLOAT = 4

BUTTON_D7 = 7
BUTTON_D4 = 4
BUTTON_D2 = 2
PUMP_PIN = 6
VALVE_PIN = 5

PULL_UP = 1
INPUT = 0

HIGH = 1
LOW  = 0

ABSOLUTE = 0
RELATIVE = 1
PATH_LINEAR = 0   # path based on linear interpolation
PATH_ANGLES = 1   # path based on interpolation of servo angles

# interpolation types
INTERP_EASE_INOUT_CUBIC = 0  # original cubic ease in/out
INTERP_LINEAR           = 1
INTERP_EASE_INOUT       = 2  # quadratic easing methods
INTERP_EASE_IN          = 3
INTERP_EASE_OUT         = 4


def getOne7BitBytesFloatArray(val):
    """
    :param val:

    """
    return bytearray([val])


def getTwo7BitBytesIntegerArray(val):
    """

    :param val:

    """
    int_val = int(val)
    return bytearray([abs(int_val) % 128, abs(int_val) >> 7])


def getThree7BitBytesFloatArray(val):
    """

    :param val:

    """
    int_val = int(val)
    decimal_val = int(round((val - int_val) * 100))
    return bytearray([abs(int_val) % 128, abs(int_val) >> 7, abs(decimal_val)])


def getThree7BitBytesIntegerArray(val):
    """

    :param val:

    """
    return bytearray([0 if val > 0 else 1, abs(val) % 128, abs(val) >> 7])


def getFour7BitBytesFloatArray(val):
    """

    :param val: Float Type
    :return: 7BitBytes Array, length = 4
    """
    int_val = int(val)
    decimal_val = int(round((val - int_val) * 100))
    return bytearray([0 if val > 0 else 1, abs(int_val) % 128, abs(int_val) >> 7, abs(decimal_val)])


def getFloatFromFour7BitBytes(array):
    """

    :param array: 4 bytes array
    :return: Float Type
    """
    negative = -1 if ord(array[0]) == 1 else 1
    val = negative * (ord(array[1]) + ord(array[2]) * 128 + ord(array[3]) / 100.00)
    return val

class Interfaceinitializationerror(Exception):
    """
    .. UnknownFirmwareException:
    """
    def __init__(self, error):
        self.error = error

    def __str__(self):
        return repr(self.error)


class NoUArmPortException(IOError):
    """
    .. _NoUArmPortException:
    """
    def __init__(self, error):
        self.error = error

    def __str__(self):
        return repr(self.error)


class UnknownFirmwareException(Exception):
    """
    .. UnknownFirmwareException:
    """
    def __init__(self, error):
        self.error = error

    def __str__(self):
        return repr(self.error)


class UnSupportedFirmwareVersionException(Exception):
    """
    .. UnSupportFirmwareVersionException:
    """
    def __init__(self, error):
        self.error = error

    def __str__(self):
        return repr(self.error)
